File: TypeScript/queen-attack.ts
import { Grid } from "./helper/grid";

type Position = readonly [number, number]

type Positions = {
  white: Position
  black: Position
}

export class QueenAttack {
  public readonly black: Position;
  public readonly white: Position;
  private board: Grid<string>;

  constructor({white, black}: Partial<Positions> = {}) {
    this.white = white ? white : [7,3];
    this.black = black ? black : [0,3];

    if([...this.black, ...this.white].filter(pos => pos >= 0 && pos < 8).length != 4) {
        throw new Error("Queen must be placed on the board");
    }

    if(JSON.stringify(this.white) === JSON.stringify(this.black)) {
        throw new Error('Queens cannot share the same space');
    }

    this.board = new Grid<string>(8, 8, '_');
    this.board.fillCell(this.white[0], this.white[1], "W");
    this.board.fillCell(this.black[0], this.black[1], "B");
  }

  toString(): string {
    return this.board.toString();
  }

  get canAttack(): boolean {
    let result = this.board.cardinalsAndOrdinals(...this.white);
    return result.flatMap(obj => obj.values).includes("B");
  }
}

let qa = new QueenAttack({ white: [1, 2], black: [5, 6] });

console.log(qa.toString());
console.log(`Can attack: ${qa.canAttack}`);

File: C#/Queen Attack/QueenAttack.cs
using _Helper;
using System;
using System.Collections.Generic;

public static class QueenAttack
{
    public static bool CanAttack(Queen white, Queen black)
    {
        var board = new _Helper.Grid<char>(8, 8, '_');

        board.FillCell(white.Row, white.Column, 'W');
        board.FillCell(black.Row, black.Column, 'B');

        Console.WriteLine(board.ToString());

        var possibleCells = new List<char>();

        possibleCells.AddRange(board.Cardinals(white.Row, white.Column));
        possibleCells.AddRange(board.Ordinals(white.Row, white.Column));

        return possibleCells.Contains('B');
    }

    public static Queen Create(int row, int column)
    {
        if (row < 0 || row > 7 || column < 0 || column > 7)
        {
            throw new ArgumentOutOfRangeException("Queen position has to be between 0-7 inclusive.");
        }

        return new Queen(row, column);
    }
}